src/java/org/apache/cassandra/auth/MutualTlsInternodeAuthenticator.java

```diff
@@ -0,0 +1,186 @@
+    private final MutualTlsCertificateValidator certificateValidator;
+    private final List<String> validIdentities;
+
+    public MutualTlsInternodeAuthenticator(Map<String, String> parameters)
+    {
+        String certificateValidatorClassName = parameters.get(VALIDATOR_CLASS_NAME);
+        if (StringUtils.isEmpty(certificateValidatorClassName))
+        {
+            String message = "internode_authenticator.parameters.validator_class_name is not set";
+            logger.error(message);
+            throw new ConfigurationException(message);
+        }
+
+        certificateValidator = ParameterizedClass.newInstance(new ParameterizedClass(certificateValidatorClassName),
+                                                              Arrays.asList("", AuthConfig.class.getPackage().getName()));
+        Config config = DatabaseDescriptor.getRawConfig();
+        checkInternodeMtlsConfigurationIsValid(config);
+        validIdentities = getIdentitiesFromKeyStore(config.server_encryption_options.outbound_keystore,
+                                                                  config.server_encryption_options.outbound_keystore_password,
+                                                                  config.server_encryption_options.store_type);
+
+        if (!validIdentities.isEmpty())
+        {
+            logger.info("Initializing internode authenticator with identities {}", validIdentities);
+        }
+        else
+        {
+            String message = String.format("No identity was extracted from the outbound keystore '%s'", config.server_encryption_options.outbound_keystore);
+            logger.info(message);
+            throw new ConfigurationException(message);
+        }
+    }
+
+    @Override
+    public boolean authenticate(InetAddress remoteAddress, int remotePort)
+    {
+        throw new UnsupportedOperationException("mTLS Authenticator only supports certificate based authenticate method");
+    }
+
+    @Override
+    public boolean authenticate(InetAddress remoteAddress, int remotePort, Certificate[] certificates, InternodeConnectionDirection connectionType)
+    {
+        return authenticateInternodeWithMtls(remoteAddress, remotePort, certificates, connectionType);
+    }
+
+
+    @Override
+    public void validateConfiguration() throws ConfigurationException
+    {
+
+    }
+
+    protected boolean authenticateInternodeWithMtls(InetAddress remoteAddress, int remotePort, Certificate[] certificates,
+                                                    IInternodeAuthenticator.InternodeConnectionDirection connectionType)
+    {
+        if (connectionType == IInternodeAuthenticator.InternodeConnectionDirection.INBOUND)
+        {
+            String identity = certificateValidator.identity(certificates);
+            if (!certificateValidator.isValidCertificate(certificates))
+            {
+                noSpamLogger.error("Not a valid outbound certificate {}", identity);
```

Comment:
We are authenticating for inbound connection and validating an outbound certificate. So I think the log message should have outbound certificate.

Reply:
The problem is outbound is ambiguous as to the type of certificate or type of connection.  This is an inbound connection by the `if` statement above - but is handling another nodes outbound certificate.

Perhaps it would be more valuable to say something like `noSpamLogger.error("Not a valid certificate from {}:{} with identity '{}'", remoteAddress, remotePort, identity);"`  then it's clear where the cert is from and who the identity is.

Reply:
That's a more clearer message. Updated the log!